README.md

```diff
@@ -62,12 +62,22 @@ Must be a list of *tensor specification keys*.
   *tensor specification keys*:
   - `name` tensor name
   - `data_type` data type (e.g. float32)
-  - `data_range` tuple of (minimum, maximum)
-  - `axes` string of axes identifying characters from: btczyx
+  - `data_range` tuple of (minimum, maximum) (e.g. [-inf, inf])
+  - `axes` string of axes identifying characters from: btczyx. Same length and order as the axes in `shape`.
   - `shape` specification of tensor shape\
-    Either as *exact shape with same length as `axes`*,\
-    or (only for input) as {`min` *minimum shape with same length as `axes`*, `step` *minimum shape change with same length as `axes`*},\
-    or (only for output) as {`reference_input` *input tensor name*, `scale` *list of factors 'output_pix/input_pix' for each dimension*, `offset` *position of origin wrt to input*, `halo` *the halo to crop from the output tensor (for example to crop away boundary efects or for tiling)*}
+    Either as *fixed shape with same length as `axes`* (e.g. `shape`: [1, 512, 512, 1]),\
+    or (only for input) as\
+    `shape`
+    - `min` *minimum shape with same length as `axes`*
+    - `step` *minimum shape change with same length as `axes`*\
+    
+    or (only for output) as\
+    `shape`
+    - `reference_input` *input tensor name*
+    - `scale` *list of factors 'output_pix/input_pix' for each dimension*
+    - `offset` *position of origin wrt to input*
+    - `halo` *the halo to crop from the output tensor* (for example to crop away boundary effects or for tiling) (e.g. the valid domain of an output of `size = [256, 256]` and `halo = [32, 32]`, results in `output_shape = [256 - 2 * 32, 256 - 2 * 32] = [192, 192]`).\
```

Comment:
In my perception the `halo` border is actually not cropped away, but returned. However, your description of `halo` makes more sense. We should never return anything invalid. I saw a problem with 'same' convolutional models, that do return the same output size but have a certain halo of (more or less) invalid border data. I suppose those models should incorporate a postprocessing crop and only return valid data, turning the same convolution model into a valid one. @constantinpape ,@m-novikov, what do you think?

Reply:
> In my perception the `halo` border is actually not cropped away, but returned.

But what's the difference then of having a halo or not?

The way I understand halo here is that it's a hint to the consumer runing prediction with this model, that indicates which "outer region" of the prediction should be discarded, for example due to boundary artifacts caused by same convolutions. (but there could be lots of other reasons for doing this)

Reply:
> But what's the difference then of having a halo or not?

I am confused, as to me you answer this question yourself:

>The way I understand halo here is that it's a hint to the consumer running prediction with this model, that indicates which "outer region" of the prediction should be discarded, for example due to boundary artifacts caused by same convolutions. (but there could be lots of other reasons for doing this)

This is how I understood it, too. I have to say though, that I like @esgomezm definition more that forces you to crop the outputs that contain a halo (either through valid convs or postprocessing), to instead return only valid output. Every consumer software should just tile accordingly and would no longer have to deal with the two options of 1. I ignore the halo and look at ugly border artefacts or 2. I crop the halo from what I got back from the runner. It would amount to option 3: I have to tile correctly (overlapping if model has halo), but what I get back is all valid :-)

Reply:
Ah ok, so now I understand the issue here, I think.
I think we all agree on the principle definition of halo, but the question is who's responsible for cropping it. 
1. It's cropped in the mode already.
2. It's cropped by the consumer.

I don't think option 1 makes so much sense, because then we wouldn't need to specify the halo in the first place, because we can just use the predictions from the model in the first place.

But if we go with option 2, this implies that we don't change the output shape when a halo is specified.

Reply:
> I don't think option 1 makes so much sense, because then we wouldn't need to specify the halo in the first place, because we can just use the predictions from the model in the first place.

right, if we do want to handle it this way, we could just move it into output `shape`, as in this case it's only applicable when calculating the actual output from a reference input